### ISSUE TITLE
ENH: Collapse some code blocks

### ISSUE THREAD

**Issue by larsoner** [sphinx-gallery/sphinx-gallery#953]
Side-remark: It would be really cool if one could specify that individual code sections can be/are folded (or maybe completely omitted?) in sphinx-gallery. This is one of the gallery entries in which only the generated figure is relevant not the code that is generating it.

_Originally posted by @timhoffm in https://github.com/matplotlib/matplotlib/issues/22912#issuecomment-1110293857_

**Comment by timhoffm** [comment]
Probably the simplest approach is adding a [`<details>`](https://developer.mozilla.org/de/docs/Web/HTML/Element/details) tag.

### closed:
<img width="583" alt="image" src="https://user-images.githubusercontent.com/2836374/165510882-df7e1244-27e9-4629-a6ba-d3acaaafea5e.png">

### open:
<img width="573" alt="image" src="https://user-images.githubusercontent.com/2836374/165511172-ce7fafe6-9356-443d-9565-5e328005c45c.png">


However, styling is not ideal (not sure how far you can improve that with CSS), and I don't think you can change the summary based on collapse state.

**Comment by larsoner** [comment]
Given that this is something that should ideally be used sparingly, maybe a block-level config that you could use would be good enough like:
```
# sphinx_gallery_nest_details: Code
```
Where whatever is after the `:` says what the summary should be. And we only wrap a given block in `details` with a `summary` if this special block-level config shows up.

Another option would be to vendor/add https://github.com/scopatz/hiddencode/blob/master/hidden_code_block.py and then we have full control over how the HTML gets rendered. We could maybe put it together with some variant/derivative/sibling of [sphinx-copybutton](https://sphinx-copybutton.readthedocs.io/en/latest/) to have the toggle something inline for example. Then a block-level config like `# sphinx_gallery_collapse` would tell it to style the code block with the `collapse` tag, and all other code blocks get collapsibility (but are un-collapsed by default) which might be nice actually. But this is a lot harder...

**Comment by timhoffm** [comment]
I totally agree with your assessment that inlined collapse buttons would be nicer, but are a lot harder.

What do you think: Would it be possible to start simple and possibly switch out the implementation with something more fancy later? Or would that too breaking for then existing users?

**Comment by alexisthual** [comment]
Hi sorry I'm a bit late in the game!
It's just my opinion, but I usually don't really like collapsing blocks ; it usually feels very clicky to me, and it makes the page's state hard to share (as we probably won't track which blocks are toggled). I think a good option would be to have a toggle/untoggle sticky button which (un)collapses all blocks.

**Comment by timhoffm** [comment]
Global toggling would not help my use case: There are two kinds of code blocks in examples, one that contains actual relevant code that one want's to communicate to the reader, and one that contains only supporting code to make the example execute. I'd like to be able to hide supporting code block blocks that are more distracting than helpful, on a case-by-case basis.

**Comment by smarie** [comment]
This feature is very interesting to me !
The most straightforward use case is a project using sphinx gallery to present some benchmark results. The code to construct all the plots (typically using plotly) is very verbose, and probably the reader does not care much about it : the reader is interested in the charts. In this situation, having the possibility to define foldable code blocks with a default state "folded" would make a lot of sense.

**Comment by larsoner** [comment]
@smarie I think there's enough support / need for this that you could take a stab at an implementation if you want

**Comment by smarie** [comment]
Thanks @larsoner ! Unfortunately I don't have the general bandwidth for such additional contributions these days - my last attempt to find some time for contribution was in #930 and did not happen yet for lack of available time :(. Just maintaining my own packages is a struggle.
Let's see if someone is willing to take this in the near future - if so I'll definitely be following this. If this does not happen, I might as well end up contributing or finding contributors
Thanks again!

**Comment by Deathn0t** [comment]
Hello, I was also looking for this functionality and I came up with the following code patch to be added inside the `conf.py`. 

It is probably not the neatest solution but it allows to access all the features of this `dropdown` directive, and it also allows it to not impact the notebook/py build (the code sections that are in the dropdown will remain present).

```python
# Patch Sphinx Gallery
import re
import sphinx_gallery.gen_rst
from textwrap import indent

def parse_dropdown(rst_text):
    pattern = re.compile(
        r"# \.\. dropdown::(.*?)\n"  # Match the title
        r"(?:    :(\w+):(?: (.*?))?\n)*",  # Match optional keys and values
        re.DOTALL
    )

    match = pattern.search(rst_text)
    if match:

        title = match.group(1).strip()

        # Extract options as key-value pairs
        options = dict(re.findall(r"    :(\w+):(?: (.*))?", rst_text))

        # Remove matched lines from rst_text
        cleaned_rst_text = pattern.sub("", rst_text)
        for _ in options:
            i = cleaned_rst_text.index("\n")
            cleaned_rst_text = cleaned_rst_text[i+1:]

        return {"title": title, "options": options}, cleaned_rst_text
    
    return None, rst_text


def codestr2rst(codestr, lang="python", lineno=None):
    """Return reStructuredText code block from code string."""

    # Start by checking if there is a dropdown directive
    dropdown_config, codestr = parse_dropdown(codestr)

    if lineno is not None:
        # Sphinx only starts numbering from the first non-empty line.
        blank_lines = codestr.count("\n", 0, -len(codestr.lstrip()))
        lineno = f"   :lineno-start: {lineno + blank_lines}\n"
    else:
        lineno = ""
    # If the whole block is indented, prevent Sphinx from removing too much whitespace
    dedent = "   :dedent: 1\n"
    for line in codestr.splitlines():
        if line and not line.startswith((" ", "\t")):
            dedent = ""
            break
    code_directive = f".. code-block:: {lang}\n{dedent}{lineno}\n"
    indented_block = indent(codestr, " " * 4)
    block = code_directive + indented_block

    # Process the dropdown configuration
    if dropdown_config is not None:
        dropdown_directive = f".. dropdown::"
        if len(dropdown_config["title"]) > 0:
            dropdown_directive += f" {dropdown_config['title']}"
        dropdown_directive += "\n"

        for key, value in dropdown_config["options"].items():
            dropdown_directive += " " * 4
            if len(value) > 0:
                dropdown_directive += f":{key}: {value}"
            else:
                dropdown_directive += f":{key}:"
            dropdown_directive += "\n"

        dropdown_directive += "\n"
        code_block = indent(block, " " * 4)
        block = dropdown_directive + code_block

    return block

sphinx_gallery.gen_rst.codestr2rst = codestr2rst
```

This works with `sphinx-design` and the `dropdown` directive ([Sphinx Design Documentation](https://sphinx-design.readthedocs.io/en/sbt-theme/dropdowns.html)).

It allows you to write the following comments in the source `*.py` of the gallery:

```python
# .. dropdown:: Imports
#    :open:
from deephyper.analysis._matplotlib import update_matplotlib_rc
from deephyper.hpo import HpProblem
from deephyper.hpo import ExperimentalDesignSearch
import matplotlib.pyplot as plt
```

and then will become:

```rst
.. dropdown:: Imports
    :open:

    .. code-block:: Python

        from deephyper.analysis._matplotlib import update_matplotlib_rc
        from deephyper.hpo import HpProblem
        from deephyper.hpo import ExperimentalDesignSearch
        import matplotlib.pyplot as plt

        update_matplotlib_rc()
```

It looks like this when it is open:

![Image](https://github.com/user-attachments/assets/c7969662-43fd-4b77-8520-db4e278df417)

and when it is closed:
![Image](https://github.com/user-attachments/assets/8cd364b8-a512-4601-8e12-71da64691be7)

**Comment by timhoffm** [comment]
Note: Sphinx has recently added collapsible support to admonitions https://github.com/sphinx-doc/sphinx/pull/12507.

This may make collapsible code blocks simpler. I'm thinking along the lines of

```
.. admonition:: Code
   :collapsible:
   :class: collapsible-code-block

   .. code-block::

      [the code]
```

plus some CSS to make margins go away and fill the whole admonition with the code.
